Implementation of EIP-1153: Transient Storage using Disk Persistence and Lifecycle Management

snissn opened the pull request:
### **Description**

This WIP PR introduces transient storage support in the Filecoin EVM, conceptually aligning with Ethereum's EIP-1153. Transient storage functionality is implemented with a tombstone mechanism for lifecycle tracking keyed by transaction identifiers (`origin` and `nonce`). While the core implementation is in place, work remains on validating the transient data's lifecycle and completing associated tests.

---

#### **Current Progress**

**State Modifications:**
1. **Persistent Transient Data:**
   - A `Cid` variable has been added to represent transient storage.

   ```rust
   pub transient_state: Cid
   ```

2. **Lifecycle Tracking:**
   - Introduced `TransientDataLifespan` to track lifecycle with origin `ActorID` and `nonce`.

   ```rust
   pub struct TransientDataLifespan {
       pub origin: ActorID,
       pub nonce: u64,
   }
   ```

**New Operations:**
1. `TLOAD`: Retrieves transient data while ensuring lifecycle validity.
2. `TSTORE`: Updates transient data and lifecycle.
3. Lifecycle validation (`TODO`): Compare transaction metadata with stored tombstone.

---

#### **Remaining TODOs**

1. **Liveliness Check for Transient Data:**
   - Validate that transient data is only accessible within the valid transaction context by comparing the tombstone (stored origin and nonce) with the current transaction.

   **Code Location:**
   - `get_transient_storage` and `set_transient_storage` in `actors/evm/src/interpreter/system.rs`.

   ```rust
   // TODO check tombstone for liveliness of data
   ```

2. **Test Coverage:**
   - Write tests to verify:
     - Transient data lifecycle (e.g., expiration between transactions).
     - Liveliness checks during nested and delegate calls.
     - Behavior when accessing invalid or expired transient data.
   - Existing transient storage operations (`TLOAD`, `TSTORE`) are partially tested.

   **Code Location:**
   - `actors/evm/src/interpreter/instructions/storage.rs` (Test module)

   ```rust
   // TODO test transient storage lifecycle
   ```

3. **Edge Case Handling:**
   - Ensure transient data is cleared appropriately in all scenarios, including contract self-destruct or re-creation.

---

#### **Testing**

**Implemented Tests:**
- `test_tload`: Validates `TLOAD` operation when data exists.
- `test_tload_oob`: Ensures out-of-bounds `TLOAD` returns zero.
- `test_tstore`: Confirms `TSTORE` updates transient data correctly.

**Pending Tests:**
- Liveliness and lifecycle tests (`TODO` in the test module).
- System-wide validation tests for lifecycle management.

---

#### **Tradeoffs and Considerations**

1. **Increased Storage Overhead:**
   - Adds `transient_state` and `transient_data_lifespan` to `State`.

2. **Gas Inefficiency:**
   - Current approach persists transient data, diverging from EIP-1153's ephemeral storage design.

---

#### **Checklist**

- [x] Core implementation of `TLOAD` and `TSTORE`.
- [x] Integration with `State` and `System`.
- [x] Basic tests for `TLOAD` and `TSTORE`.
- [x] Lifecycle validation for transient data (**TODO**).
- [x] Comprehensive test coverage (**TODO**).

---

#### **Next Steps**

- Provision work for go-state-types evm state upgrade and migration for next network upgrade started here: https://github.com/filecoin-project/go-state-types/compare/mikers/evm-transient-data-nv17?expand=1
---

#### **References**

- **EIP-1153: Transient Storage:** [Link to EIP-1153](https://eips.ethereum.org/EIPS/eip-1153)
- **Related Discussions:** https://github.com/filecoin-project/FIPs/discussions/855#discussioncomment-11117042

---

#### **Review Feedback**

Incorporates feedback from @Stebalien:
- Minimize unnecessary writes when clearing transient data.
- Keep `State` clean unless other updates occur.

Stebalien commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -91,6 +91,11 @@ pub struct System<'r, RT: Runtime> {
+
+    /// The contract's EVM transient storage slots.
+    transient_slots: StateKamt<RT::Blockstore>,
+    pub(crate) transient_data_lifespan: Option<TransientDataLifespan>,
```
Does this need `pub(crate)`?

Stebalien replied:
(I'd actually like to fix the visibility of the other cases as well, but they're a bit trickier.

snissn replied:
a98c0585 remove pub visibility from transient_data_lifespan. TODO still is combine transient_slots with transient_data_lifespan anyway

snissn commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -182,6 +191,13 @@ impl<'r, RT: Runtime> System<'r, RT> {
+                transient_store,
+                KAMT_CONFIG.clone(),
+            )
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "state not in blockstore")?,
```
ef679de1 updated

rvagg commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -385,6 +410,89 @@ impl<'r, RT: Runtime> System<'r, RT> {
+        // Reinitialize the transient_slots with a fresh KAMT
+        //let transient_store = self.rt.store().clone();
+        //self.transient_slots = StateKamt::new_with_config(transient_store, KAMT_CONFIG.clone());
+        // TODO XXX reinitialize does not currently work due to blockstore reference issues
```
I'd be interested to know what these errors are and how you end up diagnosing the problem, for my own educational purposes because your commented code looks like it should work.

Stebalien replied:
It's because `self.rt.store()` doesn't implement clone, while `&...` does, so rust is helpfully creating a reference for you.

Note: the `new` function explicitly _requires_ that the store be clonable.

Options are:

1. Lift that requirement up to the `impl` level.
2. Use `self.transient_slots.into_store()` to "take" the existing store.

snissn replied:
@Stebalien I've tried a few different variations of the "taking" option that you are suggesting. I didn't quite get anything to compile however.

Could you share a deeper explanation on the two options to either lift the requirement to the `impl` level (I'm not sure what you mean by this; which requirement? which `impl`?) and also if you can help me with the second option about taking the ownership of the datastore in order to satisfy the requirements of new_with_config. I'm still learning the ins and outs of this in rust, and would be very grateful. I would definitely not mind a code commit either if it's a simple lift for you. Thanks in advanced!

Stebalien replied:
Ah, so, this is going to be a _bit_ of a pain. The issue is that you can't leave `transient_slots` uninitialized so you need something to put there while you construct the new `Kamt`. E.g., you can have `Option<StateKamt<...>>` and put none there while you replace it.

But... doing that will also help facilitate lazy loading (you can leave it as `None` until the user first tries to load a transient slot). So it's not the end of the world.

snissn replied:
Thanks! I'm now returning from the thanksgiving break and will look into everything this week.

snissn replied:
I have learned many things since starting this :) I see that my previous attempt was very much against the RAII ideas that rust enforces. I didn't realize that was the issue. We don't need to destroy and create a new KAMT object but simply need to clear the one we have.

As a placeholder I have implemented a [loop over keys and delete implementation](https://github.com/filecoin-project/builtin-actors/pull/1588/files#diff-4c5e4db1ab6bf6de3f25e9c323a6d293b9acad0853f529377f7e54fafc986bfbR506) of `clear` for now until an efficient `clear()` can be used.  I opened a PR https://github.com/filecoin-project/ref-fvm/pull/2092 for an efficient implementation of `clear` in KAMT that I hope can be approved and it will be easy for me from there to upgrade kamt in builtin-actors to the future version that supports clear.

snissn commented:
Remaining issues: 
- [ ] Upgrade KAMT to support a `clear()` method that supports efficient resetting of KAMT object
- [ ] Comprehensive test coverage of reentry and nested contracts. Confirm if ref fvm MockRuntime tests can support delegate calls or if that test will only be able to be done at a higher abstraction like Lotus
- [ ] Ensure lazy load on first use

BigLep commented:
2024-12-04 conversation between FilB and FilOz:
* For testing, should be able to do recursive calls.
* Look at the fevm-test-kit for some inspiration.
* Test this in builtin-actors (better testing framework) (look at integration-tests .. evm-tests)
  * @snissn found an example test to hook on to. 
* Test this in lotus as well.
* migrations
  * a migration does need to get written but it is straightforward w
* discussion on `clear` and whether it's needed here

rvagg commented:
I did a quick look for a "simple" state migration that might be similar to this; I haven't found a great example yet but part of the v9 migration did introduce a couple of fields for FIP-0029 to the miner actor, `BeneficiaryTerm` and `PendingBeneficiaryTerm` onto a miner's `Info`, which were initialised with empty values. Unfortunately it's tangled up with a bunch of other complicated migration things, but I think the idea is present there:

* [top.go needed a call to `newMinerMigrator` which did this work](https://github.com/filecoin-project/go-state-types/blob/af8abe10f6c12db06a4cca92849f7ee674c5defb/builtin/v9/migration/top.go#L220)
* [among other things, that migrator took the `miner8.State` object, copied it to a `miner9.State` along with the empty fields, and saved that](https://github.com/filecoin-project/go-state-types/blob/af8abe10f6c12db06a4cca92849f7ee674c5defb/builtin/v9/migration/miner.go#L133-L181)

snissn commented:
following up from a sync with @rvagg @BigLep and @Stebalien with my own copy of notes:

1. I will look into `integration_tests/src/tests/evm_test.rs` for remaining testing needs

2. clear() isn't needed and the best way to implement this

Instead the determination of transient data validity should happen when the actor state is loaded 

State - should contain the (lifespan and kamt root) or be null

System - only needs a valid KAMT root or null

- tload checks for null and returns 0 early if null. otherwise is reads from KAMT without a need to check any lifecycle info
- tstore checks null, and if null initializes a new empty KAMT. from there it saves and sets the saved_state_root to null

3. a migration will be needed but it's a simple one

snissn commented:
Hey @Stebalien! 👋

I’ve got the PR to a point where I’d love your review again. The tests have been fleshed out to cover the completeness we discussed — specifically, reentry and nested contracts are being tested in the `integration_tests` section.

I’ve implemented most of the changes we wanted for `system.rs`, but I’m running into one issue I could use some guidance on. I’m having trouble getting lazy instantiation of the `transient_slots` KAMT to work. The sticking point is in the `set_transient_storage` method — when `self.transient_slots` is `None`, I’m not sure how best to initialize a new KAMT in that context.

Thanks in advance for your help! If you think it’d be easier to hash this out over a quick video call, I’m totally open to that — there’s quite a bit of nuance here.

Stebalien commented:
Hm. Yeah, you need a place to store the blockstore. It's doable but... at this point, I'm not actually sure if it's worth it: you'll only pay for this if your contract has used transient slots somewhere in the current transaction.

Stebalien commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -311,6 +350,16 @@ impl<'r, RT: Runtime> System<'r, RT> {
             .context_code(ExitCode::USR_SERIALIZATION, "failed to decode state")?
             .context_code(ExitCode::USR_ILLEGAL_STATE, "state not in blockstore")?;
 
+        let transient_data_lifespan = state.transient_data.unwrap().transient_data_lifespan;
```
This will panic if there are no transient slots.

snissn replied:
removing the options from transient_data_state and transient_data_lifespan helped, but also some extra checks needed to be added  https://github.com/filecoin-project/builtin-actors/commit/978b49d1e0f1e50fbecf020cadcfb1b5b8c4275a

Stebalien commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -385,6 +434,49 @@ impl<'r, RT: Runtime> System<'r, RT> {
     }
 
+    /// Returns the current transient data lifespan based on the execution environment.
+    pub fn get_current_transient_data_lifespan(rt: &RT) -> Option<TransientDataLifespan> {
```
This should just be a free function.

snissn replied:
I added that here a47c21e36338f087a92762a8bb3cc48acc2b8e99 but it caused some tests to fail which I updated here 4feec0fba90337dc6cffaa24837a801fe5032efc

Stebalien commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -385,6 +434,49 @@ impl<'r, RT: Runtime> System<'r, RT> {
+    pub fn get_current_transient_data_lifespan(rt: &RT) -> Option<TransientDataLifespan> {
+        match rt.message().origin().id() {
+            Ok(origin_id) => {
+                Some(TransientDataLifespan { origin: origin_id, nonce: rt.message().nonce() })
```
I'd consider storing this in the `System` object. It'll be the same over the course of the transaction.

snissn replied:
a47c21e36338f087a92762a8bb3cc48acc2b8e99

Stebalien commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -311,6 +350,16 @@ impl<'r, RT: Runtime> System<'r, RT> {
+                .set_root(&state.transient_data.unwrap().transient_data_state.unwrap())
+                .context_code(ExitCode::USR_ILLEGAL_STATE, "transient_state not in blockstore")?;
+        } else {
+            self.transient_slots.clear();
```
Note: if we get here, the transient slots _should_ already be empty. Although I guess it can't hurt to call this explicitly.

snissn replied:
removing the options from transient_data_state and transient_data_lifespan cleaned this up 978b49d1e0f1e50fbecf020cadcfb1b5b8c4275a

snissn replied:
I think it is good to add - it ensures there's a sanitation / matching of the data from the store into System

Stebalien commented on actors/evm/src/state.rs:
```diff
@@ -17,6 +17,26 @@ pub struct Tombstone {
+    /// The contract transient data state dictionary.
+    /// Transient Data State is a map of U256 -> U256 values.
+    /// KAMT<U256, U256>
+    pub transient_data_state: Option<Cid>,
+    /// The data representing the transient data lifespan
+    pub transient_data_lifespan: Option<TransientDataLifespan>,
```
No need for either of these to be options.

snissn replied:
that's helpful, thanks! 978b49d1e0f1e50fbecf020cadcfb1b5b8c4275a

Stebalien commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -91,6 +91,11 @@ pub struct System<'r, RT: Runtime> {
+
+    /// The contract's EVM transient storage slots.
+    transient_slots: StateKamt<RT::Blockstore>,
+    current_transient_data_lifespan: Option<TransientDataLifespan>,
```
```suggestion
    current_transient_data_lifespan: TransientDataLifespan,
```

There's no need for this to be an option, I think? If we fail to load state information, abort; that means there's a bug somewhere.

snissn replied:
Yes! I agree - thanks - looking into it

snissn replied:
One concern comes from get_current_transient_data_lifespan returning None - which happens at least in the test context - I'm looking into what happens if I throw an error if get_current_transient_data_lifespan returns None in load/new and see how that effects the tests

snissn replied:
I'm not sure how to handle current_transient_data_lifespan, in [new()](https://github.com/filecoin-project/builtin-actors/blob/3d5deba2a526d07fa26ab0eb5ed3ff3a3c9fd62e/actors/evm/src/interpreter/system.rs#L113) other than to allow current_transient_data_lifespan to be None. I could use a default value but I think None would be more accurate

snissn replied:
Also any reason for [tombstone to be an option](https://github.com/filecoin-project/builtin-actors/blob/3d5deba2a526d07fa26ab0eb5ed3ff3a3c9fd62e/actors/evm/src/interpreter/system.rs#L110) should apply to current_transient_data_lifespan. I think reverting to keeping current_transient_data_lifespan a option makes the most sense for now. I can't make sense of how to implement `new()` or `load()` to handle get_current_transient_data_lifespan returning None without making current_transient_data_lifespan an option. LMK if that makes sense

snissn replied:
Thanks for explaining on zoom - I updated the changes you requested. However there is a single failing test. [test_origin_ethaddr ](https://github.com/filecoin-project/builtin-actors/blob/837a003ab61a92338afec75245dc35b0b84593d7/actors/evm/src/interpreter/instructions/context.rs#L368) fails because it does not have an id associated with it.

snissn replied:
I think this fixes it - 3c6d9a7d - I'm changing the origin from type id/u64 to type Address

Stebalien replied:
The origin is always an ID address. I'd just remove that test entirely (I'm not sure why it's there).

snissn replied:
great thanks for clarifying - i reverted the commit that switched from id to address and deleted the test 037939aae6db470d3d9b18c3623a001ce0172977 and dddf19c820ff27b6877b28ba157bfc7e480d361f

snissn commented on actors/evm/src/interpreter/precompiles/evm.rs:
```diff
@@ -314,6 +314,8 @@ mod tests {
     #[test]
     fn bn_recover() {
         let rt = MockRuntime::default();
+        rt.in_call.replace(true);
```
4feec0fba90337dc6cffaa24837a801fe5032efc is where I added that. In the hash function tests like bn_recover it calls [System::create ](https://github.com/filecoin-project/builtin-actors/pull/1588/files/7a99e101d7c0e2cc5a0cd25f9fbc34c64d6f4f21#diff-82baeff92fde5dd26c2733d13be59a8878d51ab8eed5ca91e0750a56bee60509R364)
```
let mut system = System::create(&rt).unwrap();
```
which triggers a call to [get_current_transient_data_lifespan](https://github.com/filecoin-project/builtin-actors/pull/1588/files/7a99e101d7c0e2cc5a0cd25f9fbc34c64d6f4f21#diff-4c5e4db1ab6bf6de3f25e9c323a6d293b9acad0853f529377f7e54fafc986bfbR547) which calls message() in message in the test harness there's an assertion for [self.require_in_call()](https://github.com/filecoin-project/builtin-actors/blob/7a99e101d7c0e2cc5a0cd25f9fbc34c64d6f4f21/runtime/src/test_utils.rs#L859). Either we can leave the in_call.replace(true), set up the test differently or remove that assertion from the test?

Stebalien replied:
Oh, I see!

snissn commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -173,15 +183,38 @@ impl<'r, RT: Runtime> System<'r, RT> {
+        let current_transient_data_lifespan = get_current_transient_data_lifespan(rt)?;
+
+        // Handle transient storage based on the presence and lifespan of `transient_data`
+        // Handle transient storage based on the presence and lifespan of `transient_data`
```
thanks! fixed here fb8025da

snissn commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -311,6 +358,22 @@ impl<'r, RT: Runtime> System<'r, RT> {
 
+        if let Some(transient_data) = state.transient_data {
+            let state_transient_data_lifespan = transient_data.transient_data_lifespan;
+            let current_transient_data_lifespan = get_current_transient_data_lifespan(self.rt)?;
```
updated here 2c8febfc

snissn commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -446,3 +542,15 @@ impl<'r, RT: Runtime> System<'r, RT> {
 }
+
+/// Returns the current transient data lifespan based on the execution environment.
+pub fn get_current_transient_data_lifespan<RT: Runtime>(
```
updated here f1ac7791

snissn commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -446,3 +542,15 @@ impl<'r, RT: Runtime> System<'r, RT> {
+pub fn get_current_transient_data_lifespan<RT: Runtime>(
+    rt: &RT,
+) -> Result<TransientDataLifespan, ActorError> {
+    match rt.message().origin().id() {
```
7c6ba378 using unwrap here

snissn commented on actors/evm/src/interpreter/test_util.rs:
```diff
@@ -33,7 +33,7 @@ macro_rules! evm_unit_test {
         let code = vec![$($crate::evm_instruction!($inst)),*];
 
-        let mut system = System::new(&$rt, false);
+        let mut system = System::new(&$rt, false).expect("Failed to initialize System");
```
I think I updated everything necessary in this commit to change new() from Result back to returning Self a005cf77

snissn commented on integration_tests/src/tests/evm_test.rs:
```diff
@@ -753,3 +754,160 @@ pub fn evm_init_revert_data_test(v: &dyn VM) {
     assert_eq!(revert_data, expected);
 }
+
+#[allow(non_snake_case)]
```
do not need, removed here 8eed0c9a

snissn commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -173,15 +183,38 @@ impl<'r, RT: Runtime> System<'r, RT> {
         }
 
         let read_only = rt.read_only();
+        let current_transient_data_lifespan = get_current_transient_data_lifespan(rt)?;
```
I would feel very good about landing this and then collaborating with you or allowing you to do this in a followup PR.

Stebalien commented on actors/evm/src/interpreter/system.rs:
```diff
@@ -311,6 +357,20 @@ impl<'r, RT: Runtime> System<'r, RT> {
             .context_code(ExitCode::USR_SERIALIZATION, "failed to decode state")?
             .context_code(ExitCode::USR_ILLEGAL_STATE, "state not in blockstore")?;
 
+        if let Some(transient_data) = state.transient_data {
+            let state_transient_data_lifespan = transient_data.transient_data_lifespan;
+            if self.current_transient_data_lifespan == state_transient_data_lifespan {
+                self.transient_slots.set_root(&transient_data.transient_data_state).context_code(
+                    ExitCode::USR_ILLEGAL_STATE,
+                    "transient_state not in blockstore",
+                )?;
+            } else {
+                self.transient_slots.clear();
+            }
+        } else {
+            self.transient_slots.clear();
+        }
```
```suggestion
        match &state.transient_data {
            Some(TransientData { transient_data_state, transient_data_lifespan })
                if transient_data_lifespan == &self.current_transient_data_lifespan =>
            {
                self.transient_slots.set_root(transient_data_state).context_code(
                    ExitCode::USR_ILLEGAL_STATE,
                    "transient_state not in blockstore",
                )?
            }
            _ => self.transient_slots.clear(),
        }
```

(slightly more rusty way of writing this, although the current code is perfectly fine)

snissn replied:
That's neat, thanks for showing that and making the suggestion!

Stebalien reviewed (approved):
LGTM from a code perspective!

rvagg commented:
https://github.com/filecoin-project/go-state-types/compare/mikers/evm-transient-data-nv17?expand=1#diff-61808733785f19d73d06ebbd0b984828297ac80b9443c2a9071d88f1fd521b89R31-R32 `Nonce` and `TransientData` will need to be flipped in order in your go-state-types representation of this state (order matters because it's encoded as an array). But otherwise that migration looks great.

rvagg commented on actors/evm/tests/basic.rs:
```diff
@@ -203,6 +203,39 @@ fn test_push_last_byte() {
+                                                                                       // We expect this to fail because no changes are made
+    util::invoke_contract_expect_fail(&rt, &solidity_params_test_cleared);
+
+    //use a new address for our calling context
+    //This will cause the transient storage data to reset because the transient storage lifecycle value has changed
```
```suggestion
    // use a new address for our calling context; this will cause the transient storage
    // data to reset because the transient storage lifecycle value has changed
```

rvagg commented on actors/evm/tests/contracts/TransientStorageTest.sol:
```diff
@@ -0,0 +1,164 @@
+contract TransientStorageTest {
+
+    constructor(){
+	    _runTests();
```
```suggestion
        _runTests();
```

rvagg commented on actors/evm/tests/contracts/TransientStorageTest.sol:
```diff
@@ -0,0 +1,164 @@
+        require(retrievedValue == value, "TLOAD did not retrieve stored value within transaction");
+    }
+
+    function testLifecycleValidationSubsequentTransaction() public {
```
wouldn't this one be `// Test 2.2`?

rvagg commented on actors/evm/tests/contracts/TransientStorageTest.sol:
```diff
@@ -0,0 +1,164 @@
+
+    // Test 2.2: Verify nested contract independence
+    function testNestedContracts(address other) public returns (bool) {
+
```
```suggestion
```

rvagg commented on actors/evm/tests/contracts/TransientStorageTest.sol:
```diff
@@ -0,0 +1,164 @@
+        return true;
+    }
+
+        // New function to test reentry scenario
```
```suggestion
    // New function to test reentry scenario
```

snissn replied:
Ah yeah I had to refactor these a few times and lost the numbering  fun also it could use an auto fmt

rvagg reviewed (approved):
well done, particularly on the epic test to actual-changes ratio

snissn commented:
> LGTM from a code perspective!
That's great I think 

> https://github.com/filecoin-project/go-state-types/compare/mikers/evm-transient-data-nv17?expand=1#diff-61808733785f19d73d06ebbd0b984828297ac80b9443c2a9071d88f1fd521b89R31-R32 `Nonce` and `TransientData` will need to be flipped in order in your go-state-types representation of this state (order matters because it's encoded as an array). But otherwise that migration looks great.

Thank you!! That could have been so frustrating to catch later on!

snissn commented:
@rvagg thanks for the feedback on the solidity tests! I ran `forge fmt`, updated the comments and also removed no longer used solidity test files that I wrapped back into the TranisentStorageTest file at some point during development!

snissn commented:
@Stebalien thanks!!
